### PR TITLE
[New Rule] New Systemd Timer Created

### PR DESCRIPTION
**Pull request by Aegrah**
## Summary
Detects the creation of a systemd timer within any of the default systemd timer directories. Systemd timers can be used by an attacker to gain persistence, by scheduling the execution of a command or script. Similarly to cron/at, systemd timers can be set up to execute on boot time, or on a specific point in time, which allows attackers to regain access in case the connection to the infected asset was lost. 

## Query
Changed the query to a new_terms rule, to minimize FPs. Only TPs hit in the last 90 days on my linux test environments.

image

### CODE REVIEW

File: rules/linux/persistence_systemd_scheduled_timer_created.toml
Comment (brokensound77) [inline rules/linux/persistence_systemd_scheduled_timer_created.toml:69]: This is going to turn into whack-a-mole with dev environments and software. How about converting this from an exlusion-by-exception to an inclusion-by-exception rule.

Ex:

```suggestion
) and
(
  /* expand expected bad processes and parent processes */
  process.name in ("bash", "sh", "zsh",) or
  process.parent.name in ()
)
```

Comment (Aegrah) [reply rules/linux/persistence_systemd_scheduled_timer_created.toml:69]: @brokensound77 But that would cause a similar issue, if not worse, no? As an example, I assume (correct me if I'm wrong) you propose something similar to this:

image

There are however so many pre set-up "run-on-boot" or "run-on-set-time-period" processes initiated by systemd that I think logic like the the one proposed in this comment adds more noise than looking for *new* systemd .timer files (as this only occurs when manually creating a new systemd .timer file manually, or installing a service that does this).

Without excluding the ones in my screenshot above, it already triggered 90x on my minimal setup Ubuntu server, and every Linux distro, Python version or Linux Desktop version will have other processes running through systemd.

Additionally, unless we specify every single shell on every linux distro in `cat /etc/shells` (which you could also potentially wildcard by looking for process.name : *sh (to include all dash, bash, rbash, zsh, sh equivalents)) a bypass would still be rather simple by placing any file in any excluded directory (which you can't avoid because of all of the FPs). 

I also thought about sequencing the 2 (looking for the creation of a .timer file, and the execution within a set period of time), but this is also prone to false positives + you can't sequence it effectively if for example an adversary would change a .timer file instead of creating a new one, or executes on reboot which doesn't occur within the same timespan. 

What is your take on this? If I am missing something crucial I'd like to know :) Thanks!

Comment (brokensound77) [reply rules/linux/persistence_systemd_scheduled_timer_created.toml:69]: I think your original approach is a better, more fine-tuned approach

```sql
file where event.type == "creation" and
endsWith(file.name, "timer") and
file.path : (
  "/etc/systemd/system/*",
  "/usr/local/lib/systemd/system/*",
  "/lib/systemd/system/*",
  "/usr/lib/systemd/system/*",
  "/home/*/.config/systemd/user/*"
)
```

I agree that the `process` version of the rule is even harder to decipher FP vs TP. So the question I have is what would make the timer _creation_ inherently suspicious? Do you think it would be suspicious from certain UIDs or from certain processes. Including those in the original logic scopes this better than adding exclusions for every legitimate use.

I think it is a balance for all rules between exlusion-by-exception and inclusion-by-exception rule. Both have pros and cons. So my question is what logic would you include here that would make it more specific to suspicious timer updates?

Comment (brokensound77) [reply rules/linux/persistence_systemd_scheduled_timer_created.toml:69]: [This](https://github.com/elastic/detection-rules/pull/2600/files) is an example of whack-a-mole updates for tuning `exclusion-by-exception` rules.

Comment (Aegrah) [reply rules/linux/persistence_systemd_scheduled_timer_created.toml:69]: Yeah I agree that [This](https://github.com/elastic/detection-rules/pull/2600/files) is not a good base to build upon @brokensound77.

After giving this some more thought, and taking a second look at all of the data that is available within the documents related to this DR, I think the only way to not turn this into whack-a-mole is to do some whitelisting instead of blacklisting. Looking at all of the logs I have available from my different Linux servers; it's the process.name that makes it interesting to analyze or an obvious TN/FP. 

I would propose the following (given the limitations of the data that is available in the documents): the addition of several processes that indicate specific user behavior (manually creating a new timer). This allows a client for easy exception tuning while still maintaining visibility. 

Adding `process.name : ("*sh", "vi*", "nano", "touch", "cp")`. 

Other than this, I thought about making sequencing or process events, but these will all become whack-a-mole. So I think we can either add whitelisting for manual process.names or decide that the benefits of a rule like this do not outweigh the negatives (how easy it would be to by-pass and the unnecessary noise that it generates); implying it might be better to close the PR. 

I do however think that the rule addition mentioned above does have benefits, although relatively easy to bypass.

File: rules/linux/persistence_systemd_scheduled_timer_created.toml
Comment (w0rk3r) [inline rules/linux/persistence_systemd_scheduled_timer_created.toml:63]: Is this a extension?

File: rules/linux/persistence_systemd_scheduled_timer_created.toml
Comment (w0rk3r) [inline rules/linux/persistence_systemd_scheduled_timer_created.toml:18]: It would be interesting to double-check for auditbeat compatibility, and add setup instructions as needed.

Also, Endgame compatibility if possible

File: rules/linux/persistence_systemd_scheduled_timer_created.toml
Comment (w0rk3r) [inline rules/linux/persistence_systemd_scheduled_timer_created.toml:24]: We can remove this one for now and do it on a dedicated PR, or we can merge it without the Investigation Guide tag.

Other than that, you can compare to existing guides and adjust the structure accordingly

File: rules/linux/persistence_systemd_scheduled_timer_created.toml
Comment (Aegrah) [reply rules/linux/persistence_systemd_scheduled_timer_created.toml:18]: Removed auditbeat, added endgame compatibility.

File: rules/linux/persistence_systemd_scheduled_timer_created.toml
Comment (Aegrah) [reply rules/linux/persistence_systemd_scheduled_timer_created.toml:74]: Adjusted the structure in new commit.

**Comment by Aegrah** [comment]
This PR is using the new OSQuery investigation guide layout. Will have to wait until #2602 is merged before unit testing will no longer fail. In the meanwhile - investigation guide feedback is always welcome.

File: rules/linux/persistence_systemd_scheduled_timer_created.toml
Comment (eric-forte-elastic) [inline rules/linux/persistence_systemd_scheduled_timer_created.toml:64]: Missing new line at the end of the file. Otherwise looks good! :+1: 

```suggestion
value = "now-7d"

```

**Review by eric-forte-elastic — CHANGES_REQUESTED** [review]

**Pull request merged**
